### ISSUE TITLE
bug(gui-client/windows): GUI bail-out `Couldn't connect to named pipe server at `{path}`\n    1: All pipe instances are busy.`

### ISSUE THREAD

**Issue by ReactorScram** [CloudFire-LLC/cloudfire#5441]
This is documented in the user guide as a known issue. When this is closed, ensure it's removed from the user guide.

#### Describe the bug

1. The GUI is bailing out
2. The error message is wrong
3. It's not showing any error dialog even though an error happened, this is confusing to users
4. ` Restart-Service -Name FirezoneClientIpcService` fails with `Error 1061: The service cannot accept control messages at this time.`

#### To Reproduce

1. Open the resource-constrained test VM
2. Run the GUI and sign in
3. Close the GUI and immediately try to re-open it
4. The GUI will run for a bit and then bail out

Sometimes the IPC service even somehow gets into a state where no GUI is running and it still won't accept a GUI connection.

I replicated this synthetically on a dev machine like this:

1. Run the GUI normally and sign in so the computer has a token
2. Stop the GUI and IPC service
3. Re-compile the program to comment out the deep link server, which also provides single-instance for the GUI
4. Run a GUI instance and let it connect
5. Run a 2nd GUI instance. The second instance will throw this error.

#### Expected behavior

1. Expect the GUI to work if the IPC service is up
6. Expect the IPC service to always be up (not sure if it's hung or what, #5450 maybe related?)
7. Expect error messages to have correct formatting (#5445 will fix)
8. Expect a dialog if the GUI catches an error and bails out without crashing (#5454 will fix)
9. Expect Windows to be able to stop / restart the IPC service

#### Screenshots / Logs

GUI

```jsonl
{"time":"2024-06-19T16:56:00.9481957Z","target":"firezone_gui_client::client::auth","logging.googleapis.com/sourceLocation":{"file":"gui-client\\src-tauri\\src\\client\\auth.rs","line":"109"},"severity":"INFO","message":"Reloaded token from disk, starting in signed-in state."}
{"time":"2024-06-19T16:56:00.9484499Z","target":"firezone_gui_client::client::gui","logging.googleapis.com/sourceLocation":{"file":"gui-client\\src-tauri\\src\\client\\gui.rs","line":"466"},"severity":"INFO","apiUrl":"wss://api.firez.one/","message":"Starting connlib..."}
{"time":"2024-06-19T16:56:00.9484666Z","target":"firezone_gui_client::client::ipc","logging.googleapis.com/sourceLocation":{"file":"gui-client\\src-tauri\\src\\client\\ipc.rs","line":"91"},"severity":"INFO","clientPid":4596,"message":"Connecting to IPC service..."}
{"time":"2024-06-19T16:56:00.9485194Z","target":"firezone_gui_client::client::gui","logging.googleapis.com/sourceLocation":{"file":"gui-client\\src-tauri\\src\\client\\gui.rs","line":"235"},"severity":"ERROR","error":"Failed to restart session during app start\n\nCaused by:\n    0: Couldn't connect to named pipe server at `{path}`\n    1: All pipe instances are busy. (os error 231)","message":"run_controller returned an error"}
{"time":"2024-06-19T16:56:00.948641Z","target":"firezone_gui_client::client::gui","logging.googleapis.com/sourceLocation":{"file":"gui-client\\src-tauri\\src\\client\\gui.rs","line":"241"},"severity":"INFO","exitCode":"1"}
```

IPC service shows nothing after 16:50Z, not sure if it's hung up or what.

#### Platform (please complete the following information)

- _Component (i.e. macOS client / Linux client / Gateway / Admin portal):_ Windows GUI Client
- _Firezone Version (e.g. 1.0.0 or N/A):_ 1.0.9 11d76c4
- _OS and version: (e.g. Ubuntu 22.04 or N/A):_ Windows Server 2022
- _Deployment method: (e.g. Docker / Systemd / App Store or N/A):_ MSI from Github https://github.com/firezone/firezone/releases/tag/gui-client-1.0.9

#### Additional context

Still can't open About while the GUI is starting because of #5311 
Refs #5377

#### Hypotheses

- (Not replicated but feels very likely) `blocking_send` is deadlocking us when we try to shut down wintun https://github.com/firezone/firezone/blob/89bb7c2c5d7a81febdb3ad54294e7dd45851f857/rust/connlib/tunnel/src/device_channel/tun_windows.rs#L249
- (Replicated) If the IPC service hangs up somewhere and doesn't disconnect its end of the pipe, e.g. the GUI disconnects but the service hasn't touched the pipe and hasn't noticed yet, then the error can happen
- (Replicated on hardware) If the computer switches networks while the Client is signed in, then wintun may be unable to shut down.
- (Replicated, less likely) If something causes the deep link server to misbehave, a 2nd instance of the GUI could start
- (Didn't replicate) If the GUI doesn't release its end of the pipe cleanly, the server might not release its end immediately
- (Didn't replicate) If wintun gets a packet at the same time as we're shutting it down, the shutdown event might be dropped https://docs.rs/wintun/0.4.0/src/wintun/session.rs.html#127

**Comment by ReactorScram** [comment]
This message can replicate on `main` with just one GUI instance if you restart it quickly enough.
#5451 hasn't merged, so it might be related to that. 
#5426 mitigates it sometimes, if the GUI starts while the IPC service is cycling over, it will say all instances busy a couple times and then connect before the retry times out.

**Comment by jamilbk** [comment]
I believe this can reproduced by starting the GUI with no Internet connection. Or perhaps leaving Firezone running when there's no Internet for extended time.

Logs

[no-internet.zip](https://github.com/user-attachments/files/15975407/no-internet.zip)

**Comment by ReactorScram** [comment]
I'll try to replicate that but it shouldn't. Given this sequence:

- Reboot the system to clean up any zombie service issues
- IPC service auto-starts and waits for a connection
- User starts GUI
- GUI tries to connect to the IPC service

At that point nothing critical depends on Internet. The IPC service won't start connlib until after the GUI has connected

**Comment by jamilbk** [comment]
Ah I see. I didn't do much rebooting during my testing.

I wonder if #5355 causes a busy loop that prevents the IPC service from responding, causing the issue here.

**Comment by ReactorScram** [comment]
Maybe. I'll see if flooding the IPC service with `set_dns` requests causes them to queue up.

**Comment by ReactorScram** [comment]
I tried sending a bunch of `set_dns` messages this morning and it didn't seem to do anything.
I'll try this next:

- Start the GUI and make sure Firezone connects
- Change networks
- Restart the GUI

**Comment by ReactorScram** [comment]
Had it replicate yesterday, might have been on 144418f47. I can't get it to replicate now, even though I'm trying to use the exact same setup:
- Windows 11 dev laptop
- USB-C dock for Ethernet
- Swapping between Ethernet and Wi-Fi while the Client is signed in
- Pinging github.com (which resolves to 100.96.0.1, which happens to be an address on the LAN I'm on, must be something odd the ISP does)

Just can't get it to replicate.

**Comment by jamilbk** [comment]
I'm fairly confident this is fixed. Have been testing it over the last couple days with no issues. Optimistically closing!
